<a href="https://colab.research.google.com/github/emerson1000/notebooks/blob/main/Sistemas_de_colas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Sistemas de colas en el contexto industrial

###Los sistemas de colas son modelos matemáticos que se utilizan para describir la interacción entre clientes y servidores. Se utilizan en una variedad de contextos, incluyendo la industria, el comercio, los servicios y el transporte.

###En este trabajo, se analizarán dos tipos de sistemas de colas en el contexto industrial: el sistema MM1 y el sistema MG (M3)

##Sistema MM1

###El sistema MM1 es un sistema de colas en el que los clientes llegan a un servidor de forma independiente y aleatoria, con una tasa de llegada constante. El servidor tiene una capacidad infinita y puede atender a los clientes de forma instantánea.

##Elementos del sistema

###Los elementos del sistema MM1 son los siguientes:

###Clientes: Los clientes son las unidades que llegan al sistema para ser atendidas.
Servidor: El servidor es la unidad que atiende a los clientes.
Tasa de llegada: La tasa de llegada es la velocidad a la que llegan los clientes al sistema.
Tasa de servicio: La tasa de servicio es la velocidad a la que el servidor atiende a los clientes.

###Relaciones fundamentales

####Las relaciones fundamentales para el cálculo de medidas de desempeño e indicadores en el sistema MM1 son las siguientes:

Parámetro de utilización: El parámetro de utilización es la fracción del tiempo que el servidor está ocupado.

utilización = tasa_llegada / tasa_servicio

Tiempo medio de espera: El tiempo medio de espera es el tiempo que un cliente promedio espera en la cola antes de ser atendido.

tiempo_espera = 1 / (tasa_servicio - tasa_llegada)

Longitud media de la cola: La longitud media de la cola es el número promedio de clientes que están esperando en la cola.

longitud_cola = tiempo_espera * tasa_llegada



###Caso de uso MM1

Consideremos un cajero automático de un banco. Los clientes son las personas que llegan al cajero automático para realizar una transacción. El servidor es el cajero automático.

Se supone que la tasa de llegada de las personas es constante y es de 10 personas por hora. La tasa de servicio del cajero automático es también constante y es de 12 personas por hora.

In [ ]:
##Solución
import numpy as np
import matplotlib.pyplot as plt
#Importar librerías necesarias

# Parámetros del sistema MM1 para el cajero automático
lambda_mm1 = 10.0  # Tasa de llegada (personas por hora)
mu_mm1 = 12.0      # Tasa de servicio (personas por hora)

# Simulación del sistema MM1 para el cajero automático
tiempo_llegada_mm1 = np.random.exponential(1/lambda_mm1, 1000) # Genera 1000 tiempos de llegada con una distribución exponencial
tiempo_servicio_mm1 = np.random.exponential(1/mu_mm1, 1000) # Genera 1000 tiempos de servicio con una distribución exponencial
tiempo_espera_mm1 = np.maximum(0, tiempo_servicio_mm1 - tiempo_llegada_mm1)
# Calcula el tiempo de espera para cada cliente como la diferencia entre tiempo de servicio y tiempo de llegada,
# pero evita tiempos de espera negativos

# Medidas de desempeño para el sistema MM1 del cajero automático
promedio_tiempo_espera_mm1 = np.mean(tiempo_espera_mm1)
# Calcula el tiempo promedio de espera
utilizacion_mm1 = lambda_mm1 / mu_mm1
# Calcula la utilización del sistema

# Imprimir resultados y conclusiones
print(f"Tiempo de espera promedio en el cajero automático: {promedio_tiempo_espera_mm1} horas")
# Medidas de desempeño para el sistema MM1 del cajero automático
print(f"Utilización en el cajero automático: {utilizacion_mm1}")

Tiempo de espera promedio en el cajero automático: 0.033573418979360115 horas
Utilización en el cajero automático: 0.8333333333333334


Conclusión: Cada persona debería esperar en promedio 3.3 minutos para utilizar el cajero automático

El nivel de utilización del cajero es cercado al 83% del tiempo

###Caso de uso MM3 (MG)


En teoría de colas, el MM3 es un modelo de colas que describe un sistema en el que los clientes llegan a un servidor de forma independiente y aleatoria, con una tasa de llegada no constante. El servidor puede atender a un solo cliente a la vez, y el tiempo de servicio de cada cliente no es exponencialmente distribuido.

El modelo MM3 es un modelo más realista que el modelo MM1, que asume que la tasa de llegada es constante. El modelo MM3 también es más realista que el modelo M/M/1, que asume que el tiempo de servicio es constante.


Las medidas de desempeño del modelo MM3 se pueden calcular utilizando las siguientes fórmulas:

Parámetro de utilización:

utilización = tasa_llegada / tasa_servicio

Tiempo medio de espera:

tiempo_espera = 1 / (tasa_servicio - tasa_llegada)

Longitud media de la cola:

longitud_cola = tiempo_espera * tasa_llegada

El sistema MM3 es utilizado en una serie de aplicaciones, tales como:

Sistemas de atención al cliente

Sistemas de producción

Sistemas de transporte

En este caso presentaremos un caso práctico referido a una cafetería:

####Considere una cafetería. Los clientes son las personas que llegan al café para consumir dicho producto. El servidor es el vendedor de cafés.

###Se supone que la tasa de llegada de las personas es variable y depende del día de la semana y del momento del día. Por ejemplo, la tasa de llegada es mayor los fines de semana y en las horas pico. La tasa de servicio del camarero también es variable y depende de la cantidad de cafés que se tengan que servir.

###Determinaremos entonces a través del sistema MM3 el tiempo de espera promedio del café y la tasa de utilización de la misma

In [ ]:
# Importar bibliotecas necesarias
import numpy as np
import matplotlib.pyplot as plt

# Parámetros del sistema MG para el café
dias_semana = ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo']
tasa_llegada_diaria = {'Lunes': 20, 'Martes': 18, 'Miércoles': 22, 'Jueves': 20, 'Viernes': 25, 'Sábado': 30, 'Domingo': 25}
# Definir tasas de llegada diaria para cada día de la semana

# Simulación de tasas de servicio dependiendo de la cantidad de pedidos
pedidos_atendidos = np.random.randint(15, 30, size=1000)  # Simulación de 1000 pedidos atendidos
# Generar una cantidad aleatoria de pedidos atendidos para cada período simulado

# Simulación del sistema MG para el café
tiempo_llegada_mg = np.concatenate([np.random.exponential(1/tasa_llegada_diaria[dia], int(1000/7)) for dia in dias_semana])
# Generar 1000 tiempos de llegada para el café según la tasa diaria específica de cada día

tiempo_servicio_mg = np.concatenate([np.random.gamma(pedido, scale=0.5, size=int(1000/7)) for pedido in pedidos_atendidos])
# Generar 1000 tiempos de servicio para el café utilizando la distribución gamma y la cantidad de pedidos atendidos
# Se ajusta la escala de la distribución gamma para que los tiempos de servicio sean más rápidos



min_length = min(len(tiempo_llegada_mg), len(tiempo_servicio_mg))
# Calcular la longitud mínima entre los arreglos de tiempos de llegada y servicio
# Esto asegura que ambos arreglos tengan la misma longitud para evitar problemas en el análisis posterior

tiempo_llegada_mg = tiempo_llegada_mg[:min_length]
tiempo_servicio_mg = tiempo_servicio_mg[:min_length]
# Reducir la longitud de ambos arreglos al valor mínimo calculado
# Esto garantiza que ambos arreglos tengan la misma longitud y puedan ser comparados adecuadamente


tiempo_espera_mg = np.maximum(0, tiempo_servicio_mg - tiempo_llegada_mg)
# Calcular el tiempo de espera para cada cliente como la diferencia entre tiempo de servicio y tiempo de llegada,
#evitando tiempos de espera negativos

# Medidas de desempeño para el sistema MG del café
promedio_tiempo_espera_mg = np.mean(tiempo_espera_mg)
# Calcular el tiempo promedio de espera
utilizacion_mg = np.mean(pedidos_atendidos) / np.max(list(tasa_llegada_diaria.values()))
# Calcular la utilización del sistema como la relación entre la cantidad promedio de pedidos atendidos y la máxima tasa de llegada


# Imprimir resultados y conclusiones
print(f"Tiempo de espera promedio en el café: {promedio_tiempo_espera_mg} minutos")
 # Representado en minutos debido a la realidad de atención de un café
print(f"Utilización en el café: {utilizacion_mg}")
#Tasa de utilización de un café


Tiempo de espera promedio en el café: 9.539812888004608 minutos
Utilización en el café: 0.7345333333333334


El tiempo de espera promedio en este caso para degustar un café es de 9.6 minutos, la tasa de utilización es de un 73,45%

##Evaluación de los Sistemas de Espera (MM1 y MG)


###Evaluación de los Sistemas de Espera (MM1 y MG)
Sistema MM1 (Cajero Automático):
Tiempos de Espera:

El tiempo de espera promedio en el cajero automático es de aproximadamente 0.03 horas. (2 minutos por persona)
Este valor se obtiene considerando la tasa de llegada de 10 personas por hora y la tasa de servicio de 12 personas por hora.
Utilización:

La utilización del cajero automático es de 83%.
La utilización indica la eficiencia del sistema y es el cociente de la tasa de llegada y la tasa de servicio

###Sistema MG (Café):
####Tiempos de Espera:

El tiempo de espera promedio en el café es de aproximadamente 9,6 minutos.

####Utilización:

La utilización del café es de 73.45%
Esta medida indica la eficiencia del sistema y se calcula como la relación entre la cantidad promedio de pedidos atendidos y la máxima tasa de llegada.

###Análisis Comparativo de Rendimiento:
####Cajero Automático (MM1):

Aunque el tiempo de espera es relativamente mayor, la utilización es alta, indicando una eficiente utilización de los recursos.

###Café (MG):

Se ha ajustado el sistema para reflejar un tiempo de espera realista en un entorno de servicio rápido.
La utilización también es alta, indicando una eficiente atención a la demanda de pedidos.


###Evaluación Numérica de Costos Comparativos:
####Cajero Automático (MM1):

Los costos pueden estar asociados principalmente con el tiempo de espera prolongado, ya que los clientes podrían preferir una mayor eficiencia.

####Café (MG):

Los costos pueden estar asociados con la rápida rotación de clientes y la necesidad de mantener un personal eficiente.

###Conclusión e Interpretación:
Ambos sistemas tienen sus ventajas y desventajas dependiendo del contexto y las preferencias del cliente.
El cajero automático (MM1) es adecuado bajo contextos donde la eficiencia en el servicio es prioritaria.
El café (MG), con tiempos de espera en minutos, puede ser más atractivo en un entorno donde la experiencia del cliente es clave.
Es crucial considerar las preferencias y expectativas de los usuarios al evaluar y diseñar sistemas de espera, ya que la percepción del servicio puede ser tan importante como la eficiencia pura.
